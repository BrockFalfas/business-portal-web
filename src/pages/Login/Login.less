@import "../../styles/variables.less";

.Login {
  background-color: rgb(244,244,244);
  background-repeat: no-repeat;
  background-image: url(/images/loginBackground.svg);
  background-size: 800px 800px;
  background-position: center;
  width: 100%;
  height: 100vh;
  display: block;
  &__input-group-container.InputGroup {
    width: 100%;
    padding: 0;
    margin-bottom: 1rem;
    .InputGroup-label {
      color: #0188ee;
      font-weight: 700;
    }
    .InputGroup-errors {
      font-size: 0.8rem;
      padding-left: 0.2rem;
      padding-top: 0.2rem;
      color: @color-error;
    }
  }
  &__submit-btn.ant-btn {
    border-radius: 0;
    color: #fff;
    width: 100%;
    background-color: @primary-color;
    margin-top: 1.2rem;
  }
  &__form-container {
    width: 400px;
    margin: auto;
    position: absolute;
    top: 50%;
    left: 50%;
    margin-right: -50%;
    transform: translate(-50%, -50%);
  }
  &__image{
    &-container{
      position: absolute;
      bottom: -210px;
    left: 70px;
    }
    &-top{
      margin-bottom: 3rem;
    }
    &-bottom{
    }
  }
}
