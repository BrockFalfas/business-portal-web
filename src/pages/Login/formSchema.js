import * as Yup from 'yup';
import formUtils from '~utils/forms';

const formFields = {
  email: {
    label: 'E-mail',
    validator: Yup.string()
      .ensure()
      .required()
      .email()
      .min(6),
    input: {
      type: 'email',
      placeholder: 'Enter your email',
    },
  },
  password: {
    label: 'Password',
    validator: Yup.string()
      .ensure()
      .required(),
    input: {
      type: 'password',
      placeholder: 'Enter your password',
    },
  },
};

const validationSchema = formUtils.formik.makeValidationSchema(formFields);
const initialValues = formUtils.formik.makeEmptyInitialValues(formFields);

export { formFields, validationSchema, initialValues };
