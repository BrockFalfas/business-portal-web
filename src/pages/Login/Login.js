import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import { Button } from 'antd';
import FormField from '~components/FormField';

import './Login.css';

import { formFields, validationSchema, initialValues } from './formSchema';
import NotificationService from '../../services/notification';

export class Login extends React.Component {
  static propTypes = {
    login: PropTypes.func,
  };

  renderForm = ({ handleSubmit, isSubmitting, dirty }) => (
    <form onSubmit={handleSubmit}>
      {Object.entries(formFields).map(([name, options]) => {
        return (
          <FormField
            key={name}
            name={name}
            label={options.label}
            {...options.input}
            className="Login__input-group-container"
          />
        );
      })}

      <Button
        className="Login__submit-btn"
        type="submit"
        htmlType="submit"
        disabled={!dirty || isSubmitting}
        loading={isSubmitting}>
        Sign in
      </Button>
    </form>
  );

  render() {
    return (
      <div className="Login">
        <div className="Login__form-container">
          <img className="Login__image-top" src="images/loginImage.png" alt="" />
          <Formik
            initialValues={initialValues}
            onSubmit={this.handleSubmit}
            validationSchema={validationSchema}>
            {this.renderForm}
          </Formik>
          <div className="Login__image-container">
            <img className="Login__image-bottom" src="images/poweredBy.png" alt="" />
          </div>
        </div>
      </div>
    );
  }

  checkLogin = async data => {
    const { login, location } = this.props;
    const tenant = new window.URLSearchParams(location.search).get('tenant');
    const { email, password } = data;
    await login({ email, password, tenant });
  };

  handleSubmit = async (data, form) => {
    try {
      await this.checkLogin(data);
    } catch (err) {
      if (err.response && err.response.status === 401) {
        NotificationService.open({
          type: 'error',
          message: err.response.data.error,
          description: 'Try again.',
        });
      }
      form.setSubmitting(false);
    }
  };
}

const mapDispatch = ({ auth: { login } }) => ({ login });

export default connect(null, mapDispatch)(Login);
