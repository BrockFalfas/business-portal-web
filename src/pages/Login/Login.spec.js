import React from 'react';
import { shallow } from 'enzyme';

import { Login } from './Login';

jest.mock('./formSchema', () => ({
  initialValues: 'fake initial values',
  validationSchema: { cast: jest.fn(x => x) },
  formFields: {
    email: { label: 'Email' },
    password: { label: 'Password' },
  },
}));

const initLogin = overrides => {
  const mockProps = {
    login: jest.fn(),
  };
  const wrapper = shallow(<Login {...mockProps} {...overrides} />);
  return { mockProps, wrapper };
};

describe('page: Login', () => {
  it('should render without crashing', () => {
    const { wrapper } = initLogin();

    expect(wrapper).toBeTruthy();
  });

  describe('renderForm', () => {
    const { wrapper } = initLogin();
    const instance = wrapper.instance();

    it('should render without crashing', () => {
      const mockFormProps = {
        handleSubmit: jest.fn(),
        isSubmitting: false,
        values: { email: '', password: '' },
        dirty: false,
      };
      const renderedForm = shallow(instance.renderForm(mockFormProps));

      expect(renderedForm).toMatchSnapshot();
    });

    it('should disable submit button if form is not dirty or is submitting', () => {
      const mockFormProps = {
        handleSubmit: jest.fn(),
        isSubmitting: false,
        values: { email: '', password: '' },
        dirty: true,
      };

      let renderedForm = shallow(instance.renderForm(mockFormProps));
      expect(renderedForm.find('Button').prop('disabled')).toBe(false);

      renderedForm = shallow(instance.renderForm({ ...mockFormProps, dirty: false }));
      expect(renderedForm.find('Button').prop('disabled')).toBe(true);

      renderedForm = shallow(instance.renderForm({ ...mockFormProps, isSubmitting: true }));
      expect(renderedForm.find('Button').prop('disabled')).toBe(true);
    });
  });

  describe('login submit action', () => {
    it('should submit form with proper data', async () => {
      const { wrapper } = initLogin();
      const instance = wrapper.instance();
      instance.checkLogin = jest.fn();

      const fakeLoginData = {
        email: 'test@test.com',
        password: '123',
      };

      await instance.handleSubmit(fakeLoginData);

      expect(instance.checkLogin).toHaveBeenCalledWith({
        email: 'test@test.com',
        password: '123',
      });
    });
  });
});
