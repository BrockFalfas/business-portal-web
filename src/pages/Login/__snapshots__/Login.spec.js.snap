// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page: Login renderForm should render without crashing 1`] = `
<form
  onSubmit={[Function]}
>
  <FormField
    className="Login__input-group-container"
    key="email"
    label="Email"
    name="email"
  />
  <FormField
    className="Login__input-group-container"
    key="password"
    label="Password"
    name="password"
  />
  <Button
    className="Login__submit-btn"
    disabled={true}
    htmlType="submit"
    loading={false}
    type="submit"
  >
    Sign in
  </Button>
</form>
`;
